### PR TITLE
UDP protocol on connect agnhost command

### PR DESCRIPTION
**Pull request by knabben**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Adding UDP support on `agnhost connect`  subcommand, so we can have standardized tooling for e2e network probe tests. 

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:
With a host listening on UDP 8000:

```
./agnhost serve-hostname --udp --http=false --port 8000
```

```
$ ./agnhost connect --protocol udp localhost:8000
$ echo $?
0

$ ./agnhost connect --protocol udp localhost:8002  # wrong port
REFUSED: read udp [::1]:51142->[::1]:8002: read: connection refused
$ echo $?
1
```

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: test/images/agnhost/connect/connect.go
Comment (aojea) [inline test/images/agnhost/connect/connect.go:151]: we can't assume the other end knows what `hostname` means, 
this is the problem with UDP, it is connection less, so we need to send something and the other reply in order to know the connection is ok.
What if we add another flag to be able to send data  and we default it to `hostname` or `hello`
```
		cmd = []string{"/agnhost", "connect", fmt.Sprintf("%s:%d", addrTo, toPort), "--timeout=1s", "--protocol=udp","--data=hostname"}
```

File: test/images/agnhost/connect/connect.go
Comment (aojea) [inline test/images/agnhost/connect/connect.go:155]: we have to verify that we actually received some data
https://github.com/kubernetes/kubernetes/blob/bae644e85e085ce668c3edf23e8789fb623331b4/test/e2e/network/service.go#L536-L561

File: test/images/agnhost/connect/connect.go
Comment (knabben) [reply test/images/agnhost/connect/connect.go]: sure, what about `--udp-data`, since it's only used for UDP?

File: test/images/agnhost/connect/connect.go
Comment (knabben) [reply test/images/agnhost/connect/connect.go]: raising an error when `readBytes == 0`.. 
looks something that can be controlled on a flag i.e. `--wait-response`. or even use a `--udp-request` | `--udp-response`..

Comment (aojea) [reply test/images/agnhost/connect/connect.go]: if you don't receive any bytes you can not known that the other side is receiving packets.
This is the same we discussed in slack regarding netcat, it just checks that the packet goes out, it doesn't mean it has been received, there is no ACK... :shrug:

File: test/images/agnhost/connect/connect.go
Comment (aojea) [reply test/images/agnhost/connect/connect.go]: yeah, sounds much better, but keep it defaulted with some data

File: test/images/agnhost/connect/connect.go
Comment (aojea) [inline test/images/agnhost/connect/connect.go:55]: we should put some data as default, "hello" or "hostname" can be valid since are used in our tests in a number of cases, I think `hostname` is used in more places

File: test/images/agnhost/connect/connect.go
Comment (aojea) [reply test/images/agnhost/connect/connect.go]: maybe you can return an error more descriptive "NO DATA RECEIVED" , and clarifiy that we can't guarantee the port is open

File: test/images/agnhost/connect/connect.go
Comment (claudiubelu) [inline test/images/agnhost/connect/connect.go:169]: So, basically, we just have to make sure that the UDP server sends us some data back, the content of which is unimportant, right?

Comment (aojea) [reply test/images/agnhost/connect/connect.go:169]: I think this is the more general approach .... we can complicate it more and add an option to match the answer

**Comment by dims** [comment]
@knabben looks like we are adding a new "feature", please update test/images/agnhost/VERSION

**Comment by aojea** [comment]
yeah, I forget that, sorry, I think you have to bump these 3 files:

build/dependencies.yaml
test/images/agnhost/VERSION
test/images/agnhost/agnhost.go

**Comment by knabben** [comment]
I was about to ask what is the release process :) thanks @aojea @dims

**Comment by dims** [comment]
/hold

some more files need to be updated.

**Comment by knabben** [comment]
@dims bumped the files. Is this something that needs to be changed as well? https://github.com/kubernetes/kubernetes/blob/ead0271fcfbf0ee482e14151f338a6590e38e3f5/build/dependencies.yaml#L12

File: build/dependencies.yaml
Comment (dims) [inline build/dependencies.yaml:4]: Don't change this just yet (since this version does not exist in the gcr repository)

File: test/images/agnhost/agnhost.go
Comment (dims) [inline test/images/agnhost/agnhost.go:54]: Same as above

File: build/dependencies.yaml
Comment (knabben) [reply build/dependencies.yaml]: oh.. the dependencies job is failing with a `test/images/agnhost/VERSION` mismatch.

Comment (aojea) [reply build/dependencies.yaml]: @dims I think that is correct
>   # agnhost: bump this one first
there is another place that is the one that has to be bumped later, once the promotion has been done

**Comment by aojea** [comment]
:thinking: it fails 
```
hack/make-rules/../../hack/verify-vendor.sh
make: *** [Makefile:128: verify] Error 1
+ EXIT_VALUE=2
+ set +o xtrace
```
you can run it locally to find out what is the problem ...

**Comment by knabben** [comment]
locally the command pass, let me rerun this again here.

/test pull-kubernetes-dependencies

File: test/images/agnhost/agnhost.go
Comment (aojea) [reply test/images/agnhost/agnhost.go:54]: I think is correct, this has to be bumped first so we can promote it later

File: test/images/agnhost/connect/connect.go
Comment (aojea) [reply test/images/agnhost/connect/connect.go:169]: more history about this https://nmap.org/book/scan-methods-udp-scan.html#:~:text=Fortunately%2C%20Nmap%20can%20help%20inventory,packet%20to%20every%20targeted%20port.

**Comment by dims** [comment]
/hold cancel
/approve

please get an `/lgtm` from @aojea :) you main need additional approvals for the `build/` directory

**Review by spiffxp — APPROVED** [review]
/approve
/lgtm
Now that `connect` supports N=3 protocols, I think it'd be appropriate to refactor instead of continuing to copy-paste.  It feels to me like there's plenty of untested room for subtle bugs between impls, but I won't block for that.

**Comment by aojea** [comment]
> Now that `connect` supports N=3 protocols, I think it'd be appropriate to refactor instead of continuing to copy-paste. It feels to me like there's plenty of untested room for subtle bugs between impls, but I won't block for that.

yeah, but the networking tests are a minefield, and this only probes for connectivity , i.e. I submit something the other side replays something, there are other tests that need to check the content of the replay, i.e. to verify that the the pod belong to the expected hostname, so this is no a replacement for those.

**Pull request merged**
